code/modules/wiremod/components/action/pathfind.dm

```diff
@@ -0,0 +1,123 @@
+/**
+ * # Pathfinding component
+ *
+ * Calcualtes a path, returns a list of entities. Each entity is the next step in the path. Can be used with the direction component to move.
+ */
+/obj/item/circuit_component/pathfind
+	display_name = "Pathfinder"
+	display_desc = "When triggered, the next step to the target's location as an entity. This can be used with the direction component and the drone shell to make it move on its own. The Id Card input port is for considering ID access when pathing, it does not give the shell actual access."
+	circuit_flags = CIRCUIT_FLAG_INPUT_SIGNAL|CIRCUIT_FLAG_OUTPUT_SIGNAL
+
+	var/datum/port/input/input_X
+	var/datum/port/input/input_Y
+	var/datum/port/input/id_card
+
+	var/datum/port/output/output
+	var/datum/port/output/finished
+	var/datum/port/output/no_path
+	var/datum/port/output/cooldown
+	var/datum/port/output/bad_id
+
+	var/list/path
+	var/turf/old_dest
+	var/turf/next_turf
+
+	/// Cooldown to limit how frequently we can path to the same location.
+	var/same_path_cooldown = 5 SECONDS
+	var/different_path_cooldown = 30 SECONDS
+
+/obj/item/circuit_component/pathfind/get_ui_notices()
+	. = ..()
+	/// Not necessary to show the same path cooldown, since it doesn't change much for the player
+	. += create_ui_notice("Pathfinding Cooldown: [DisplayTimeText(different_path_cooldown)]", "orange", "stopwatch")
+
+/obj/item/circuit_component/pathfind/Initialize()
+	. = ..()
+	input_X = add_input_port("Target X", PORT_TYPE_NUMBER, FALSE)
+	input_Y = add_input_port("Target Y", PORT_TYPE_NUMBER, FALSE)
+	id_card = add_input_port("ID Card", PORT_TYPE_ATOM, FALSE)
+
+	output = add_output_port("Next step", PORT_TYPE_ATOM)
+	finished = add_output_port("Arrived to destination", PORT_TYPE_SIGNAL)
+	no_path = add_output_port("Failed: Can't pathfind there", PORT_TYPE_SIGNAL)
+	cooldown = add_output_port("Failed: Cooldown active", PORT_TYPE_SIGNAL)
+	bad_id = add_output_port("Failed: ID not valid", PORT_TYPE_SIGNAL)
```

Comment:
you don't need this many failed ports. You basically just need a "failed" port and a "failed message" port.

Output your reason to the Failed Message port and signal the Failed port. This is overkill